### PR TITLE
Constelación Serpens (Chall02) / String to Morse

### PR DESCRIPTION
**Pull request by eligero**
Constelación Serpens (Chall02) / String to Morse

Solved with key-value pair dictionary. The achieved solution has fits
the challenge problem's conceivement, by blacklisting wrongs behaviors.

### CODE REVIEW

**Comment by eligero** [comment]
Improved solution with classes and static variables

**Comment by eligero** [comment]
Improved solution of Constelación Serpens (Chall02) / String to Morse

Solved with key-value pair dictionary as static variable in a class.
The achieved solution has fitted challenge's problem conception, by
blacklisting wrong inputs by using regular expressions and, or, wrong
invocations. Memory clean.

**Comment by eligero** [comment]
There is no need of classes. Speed up with local variables, as they are faster than global variables on python.

File: chall02/eligero.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		'x':'-..-', 'y':'-.--', 'z':'--..'}
+	cipher = ''
+	for char in message:
+		if char >= 'a' and char <= 'z':
+			cipher += morse_dict[char]
+		elif  char >= 'A' and char <= 'Z':
+			cipher += morse_dict[char.lower()]
+		elif char == ' ':
+			cipher += ' '
+	return (cipher)
```
Comment (clafoutis42) [inline chall02/eligero.py:42]: You could have made this more pythonic with this oneliner
```suggestion
	return "".join(morse_dict.get(c, " ") for c in message)
```

File: chall02/eligero.py [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+def main():
+	if len(sys.argv) != 2 or sys.argv[1] == '' or \
+		re.search(r'[^a-zA-Z ]', sys.argv[1]):
```
Comment (clafoutis42) [inline chall02/eligero.py:46]: Using regex is a little overkill, you could have simply used the `.replace()` and `.isalpha` string methods in order to sanitize the string:
```suggestion
	if len(sys.argv) != 2 or sys.argv[1].replace(" ", "").isalpha():
```
Note that it also covers the `sys.argv[1] == ""` condition as `"".isalpha()` returns `False`

File: chall02/eligero.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+		print ('usage: ' + sys.argv[0] + ' ' + '<a-zA-Z string>')
+	else:
+		print (encrypt(sys.argv[1]))
```
Comment (clafoutis42) [inline chall02/eligero.py:49]: You could have made the code shorter by doing this
```suggestion
		return print ('usage: ' + sys.argv[0] + ' ' + '<a-zA-Z string>')
	print (encrypt(sys.argv[1]))
```

**Review by clafoutis42 — COMMENTED** [review]
Good job, you have a working solution, however it is not very pythonic. You are writing your code as if you were using C, I suggest you read [Fluent Python](https://www.amazon.com/Fluent-Python-Concise-Effective-Programming/dp/1491946008) if you want to keep progressing in python!

File: chall02/eligero.py (hunk 1 above)
Comment (eligero) [reply chall02/eligero.py:42]: That's cool. I know, but I priorized readable code. But I take note of this for future challenges.

File: chall02/eligero.py (hunk 2 above)
Comment (eligero) [reply chall02/eligero.py:46]: That's cool to know, but memory accesses & replacements are overkill in scenarios like microcontrollers were regex performs far better. That said, I take note for future challenges.

File: chall02/eligero.py (hunk 3 above)
Comment (eligero) [reply chall02/eligero.py:49]: This is a fault on my side. I take note.
Your comments on my code are welcome. Thank you.

File: chall02/eligero.py (hunk 1 above)
Comment (clafoutis42) [reply chall02/eligero.py:42]: I find my code easier to read. That said, it is my opinion, but your multiple conditions aren't easy to understand.  
A more readable solution would have been to put my one_liner in a function that has a clear name about what it does, and call it in the return close.

File: chall02/eligero.py (hunk 2 above)
Comment (clafoutis42) [reply chall02/eligero.py:46]: In this case, and many others, using regular expressions will make more memory accesses than a simple replace and isalpha.

File: chall02/eligero.py (hunk 1 above)
Comment (eligero) [reply chall02/eligero.py:42]: Those are opinions. Both are respectable.

File: chall02/eligero.py (hunk 2 above)
Comment (eligero) [reply chall02/eligero.py:46]: Of course, here is an scenario-dependent situation. That said, read memory accesses are faster and less power consumption than write ones. I suggest you read [Computer Architecture: A Quantitative Approach](https://www.amazon.es/Computer-Architecture-Quantitative-Approach-Kaufmann-ebook/dp/B078MFDTX4/ref=sr_1_1?__mk_es_ES=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1BVRNZM5TEDH2&dchild=1&keywords=computer+architecture+a+quantitative+approach&qid=1586289993&sprefix=computer+arch%2Caps%2C153&sr=8-1) if you want to get any documentation about my comments.

That said, I took note for the incomming challenges: prefer replacements over regexp.

Comment (clafoutis42) [reply chall02/eligero.py:46]: Don't forget that you're talking to a 3 year experienced python back-end developer :)  
You should really read [this book](https://www.amazon.com/Fluent-Python-Concise-Effective-Programming/dp/1491946008).

Comment (eligero) [reply chall02/eligero.py:46]: Don't forget that you're talking to a 8+ years experienced embedded systems engineer and FPGA expert :).
Yo should really read [this book](https://www.amazon.es/Computer-Architecture-Quantitative-Approach-Kaufmann-ebook/dp/B078MFDTX4/ref=sr_1_1?__mk_es_ES=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1BVRNZM5TEDH2&dchild=1&keywords=computer+architecture+a+quantitative+approach&qid=1586289993&sprefix=computer+arch%2Caps%2C153&sr=8-1).
The show must go on! :)

Comment (clafoutis42) [reply chall02/eligero.py:46]: I did a benchmark, I sent you in slack the complete results, but I'll just place this one here:
```
./pistache.py 'big gooood string aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr big gooood string aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr big gooood string aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'
3.407823218
./eligero.py 'big gooood string aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr big gooood string aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr big gooood string aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa rrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrrr'
98.57283754500001
```
From the benchmark we can tell 2 things:
- My suggestions make your solution faster than your original solution
- Your solution is only faster than mine in finding errors inside small strings (and this you can say thanks to the regex)

Comment (eligero) [reply chall02/eligero.py:46]: Of course, as I said earlier: it's my fault. Your solution should perform better, as it iterates on the dict keys first not over the all string firstly. It, obviously, should perform faster & better.

That said, Python is a interpreted language, which relies on the interpreter wich, again, relies on harware. It's perfectly possible an interpreter build for a custom hardware where my solution could perform better. Although this is possible, is not real, to my knowledge, by the way standard is x86, ARM and RISC. As I said earlier, on the real world your solution has to perform far better.

Problem here is that I was wirting about cost of read operations on memory versus write operations on memory, in which case there is no posible debate nowadays.

You should read [this book](https://www.amazon.es/dp/0385347758/ref=as_li_tl?ie=UTF8&linkCode=gs2&linkId=BSMOAZWQRFJAVPRJ&creativeASIN=0385347758&tag=why059-21&creative=390957&camp=1789).

You win. By the way, there wasn't a war on my side.

Show is ended. :)

**Pull request merged**